core/hash_map.h

```diff
@@ -74,13 +78,14 @@ class HashMapHahserDefault {
  * @param TKey  Key, search is based on it, needs to be hasheable. It is unique in this container.
  * @param TData Data, data associated with the key
  * @param Hasher Hasher object, needs to provide a valid static hash function for TKey
+ * @param Comparator comparator object, needs to be able to safely compare two TKey values
```

Comment:
It's not a workaround. Other hashmaps do this, or do a bitwise comparison of members to check for collisions. We can't do a bitwise comparison due to the way variant is implemented, thus we do a different equality check. The alternative would be to have a comparison function that internally inside of variant does a bitwise compare of the parts of variant for each type that makes sense. A separate comparator would still be needed and that comment would still be the same. 

For instance a class that has nested pointers to other class instances that it would want to compare the dereferenced value of would equally need to take care of this.

It is not possible for any hashmap to not treat nan special, or do a bitwise comparison (making sure that all different types of NaNs hash to different values also). Try inserting a NaN into a an std::map as a key for instance. (it will crash for the same reason as godot's hashmap does once you insert another item)

I can expand the comment to say that 'safely' means that x == x is guaranteed if you think that's more useful. I'm not going to add an 'intro into basic float precautions' into a comment about a hashmap.